### ISSUE TITLE
heap-buffer-overflow at find_rawstring_end  search.c:1605:24

### ISSUE THREAD

**Issue by comicfans** [neovim/neovim#11180]
- `nvim --version`:
```
NVIM v0.5.0-164-g3b3a40978
Build type: Debug
LuaJIT 2.1.0-beta3
Compilation: /home/wangxinyu/.local/bin/clang -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=always -fno-sanitize-recover=all -fno-omit-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -fsanitize=undefined -fsanitize-blacklist=/home/wangxinyu/neovim/src/.asan-blacklist -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DEXITFREE -I/home/wangxinyu/neovim/build/config -I/home/wangxinyu/neovim/src -I/home/wangxinyu/neovim/.deps/usr/include -I/usr/include -I/home/wangxinyu/neovim/build/src/nvim/auto -I/home/wangxinyu/neovim/build/include
Compiled by wangxinyu@archlinux

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/home/wangxinyu/.local/share/nvim"

Run :checkhealth for more info

```
- `vim -u DEFAULTS` (version: ) behaves differently?
```
vim -u DEFAULTS 8.1.1186 not crash
```
- Operating system/version:  archlinux 64bit
- Terminal name/version:konsole
- `$TERM`: TERM=xterm-256color

### Steps to reproduce using `nvim -u NORC`

edit a rust source, typing 

```rust
for i in 100..vector.len()-1 {}
```
and press esc to exit edit mode, then press i to let cursor between '{' and '}' ,and press enter to create new line, neovim 100% crash (but only happens if I edit particular rs file), even run with -u NORC. with some try, I've found that it crash only if have (tried) to do auto-completion.  I'm using debug version neovim so auto-completion introduces a delay during typing. if no such delay happend, neovim not crash (not 100% confirmed)

asan report 

```
=================================================================
==nvim==20641==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6210009fe903 at pc 0x00000153dd81 bp 0x7ffe14441f20 sp 0x7ffe14441f18
READ of size 1 at 0x6210009fe903 thread T0
    #0 0x153dd80 in find_rawstring_end /home/wangxinyu/neovim/build/../src/nvim/search.c:1605:24
    #1 0x15382e4 in findmatchlimit /home/wangxinyu/neovim/build/../src/nvim/search.c:1955:18
    #2 0xe8f2bc in find_start_rawstring /home/wangxinyu/neovim/build/../src/nvim/indent_c.c:121:15
    #3 0xe71e6b in get_c_indent /home/wangxinyu/neovim/build/../src/nvim/indent_c.c:1852:12
    #4 0x960ecc in f_cindent /home/wangxinyu/neovim/build/../src/nvim/eval.c:7854:28
    #5 0x8868ce in call_func /home/wangxinyu/neovim/build/../src/nvim/eval.c:6616:11
    #6 0x899407 in get_func_tv /home/wangxinyu/neovim/build/../src/nvim/eval.c:6359:11
    #7 0x926d0b in eval7 /home/wangxinyu/neovim/build/../src/nvim/eval.c:4461:15
    #8 0x92349e in eval6 /home/wangxinyu/neovim/build/../src/nvim/eval.c:4171:7
    #9 0x92101c in eval5 /home/wangxinyu/neovim/build/../src/nvim/eval.c:4019:7
    #10 0x91caf1 in eval4 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3755:7
    #11 0x91bfc7 in eval3 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3673:7
    #12 0x91b4b7 in eval2 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3604:7
    #13 0x8806fc in eval1 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3531:7
    #14 0x87fb6e in eval0 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3491:9
    #15 0x8f5dcd in ex_return /home/wangxinyu/neovim/build/../src/nvim/eval.c:23397:10
    #16 0xbce6bb in do_one_cmd /home/wangxinyu/neovim/build/../src/nvim/ex_docmd.c:2274:5
    #17 0xbafc2e in do_cmdline /home/wangxinyu/neovim/build/../src/nvim/ex_docmd.c:596:20
    #18 0x8b7cac in call_user_func /home/wangxinyu/neovim/build/../src/nvim/eval.c:23150:3
    #19 0x88625d in call_func /home/wangxinyu/neovim/build/../src/nvim/eval.c:6601:11
    #20 0x899407 in get_func_tv /home/wangxinyu/neovim/build/../src/nvim/eval.c:6359:11
    #21 0x926d0b in eval7 /home/wangxinyu/neovim/build/../src/nvim/eval.c:4461:15
    #22 0x92349e in eval6 /home/wangxinyu/neovim/build/../src/nvim/eval.c:4171:7
    #23 0x92101c in eval5 /home/wangxinyu/neovim/build/../src/nvim/eval.c:4019:7
    #24 0x91caf1 in eval4 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3755:7
    #25 0x91bfc7 in eval3 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3673:7
    #26 0x91b4b7 in eval2 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3604:7
    #27 0x8806fc in eval1 /home/wangxinyu/neovim/build/../src/nvim/eval.c:3531:7
    #28 0x881f5e in eval_to_number /home/wangxinyu/neovim/build/../src/nvim/eval.c:1185:7
    #29 0xe60153 in get_expr_indent /home/wangxinyu/neovim/build/../src/nvim/indent.c:455:17
    #30 0x826cde in fixthisline /home/wangxinyu/neovim/build/../src/nvim/edit.c:7015:18
    #31 0xe9fb91 in do_c_expr_indent /home/wangxinyu/neovim/build/../src/nvim/indent_c.c:3680:5
    #32 0x74ab13 in open_line /home/wangxinyu/neovim/build/../src/nvim/change.c:1709:5
    #33 0x84db11 in ins_eol /home/wangxinyu/neovim/build/../src/nvim/edit.c:8665:12
    #34 0x834ed6 in insert_handle_key /home/wangxinyu/neovim/build/../src/nvim/edit.c:1145:10
    #35 0x7e819c in insert_execute /home/wangxinyu/neovim/build/../src/nvim/edit.c:796:10
    #36 0x17909ed in state_enter /home/wangxinyu/neovim/build/../src/nvim/state.c:69:26
    #37 0x7efedb in insert_enter /home/wangxinyu/neovim/build/../src/nvim/edit.c:482:5
    #38 0x7e3dd9 in edit /home/wangxinyu/neovim/build/../src/nvim/edit.c:1387:3
    #39 0x11895b6 in invoke_edit /home/wangxinyu/neovim/build/../src/nvim/normal.c:7609:7
    #40 0x1150949 in nv_edit /home/wangxinyu/neovim/build/../src/nvim/normal.c:7581:5
    #41 0x113816f in normal_execute /home/wangxinyu/neovim/build/../src/nvim/normal.c:1131:3
    #42 0x17909ed in state_enter /home/wangxinyu/neovim/build/../src/nvim/state.c:69:26
    #43 0x10f2c67 in normal_enter /home/wangxinyu/neovim/build/../src/nvim/normal.c:463:3
    #44 0xeeca33 in main /home/wangxinyu/neovim/build/../src/nvim/main.c:576:3
    #45 0x7f89b9a1ece2 in __libc_start_main (/usr/lib/libc.so.6+0x23ce2)
    #46 0x42305d in _start (/home/wangxinyu/.local/bin/nvim+0x42305d)

0x6210009fe903 is located 3 bytes to the right of 4096-byte region [0x6210009fd900,0x6210009fe900)
allocated by thread T0 here:
    #0 0x4b2ed8 in malloc /home/wangxinyu/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cc:145:3
    #1 0x100bc22 in try_malloc /home/wangxinyu/neovim/build/../src/nvim/memory.c:71:15
    #2 0x100bdfc in xmalloc /home/wangxinyu/neovim/build/../src/nvim/memory.c:105:15
    #3 0xfaf749 in mf_alloc_bhdr /home/wangxinyu/neovim/build/../src/nvim/memfile.c:520:17
    #4 0xfaea8b in mf_new /home/wangxinyu/neovim/build/../src/nvim/memfile.c:255:10
    #5 0xfbf6aa in ml_new_data /home/wangxinyu/neovim/build/../src/nvim/memline.c:2826:16
    #6 0xfe6c1f in ml_append_int /home/wangxinyu/neovim/build/../src/nvim/memline.c:2129:14
    #7 0xfdccdd in ml_flush_line /home/wangxinyu/neovim/build/../src/nvim/memline.c:2807:15
    #8 0xfe228a in ml_get_buf /home/wangxinyu/neovim/build/../src/nvim/memline.c:1852:5
    #9 0x13a6b72 in reg_getline /home/wangxinyu/neovim/build/../src/nvim/regexp.c:3282:10
    #10 0x14378b9 in nfa_regexec_both /home/wangxinyu/neovim/build/../src/nvim/regexp_nfa.c:6486:12
    #11 0x13e9291 in nfa_regexec_multi /home/wangxinyu/neovim/build/../src/nvim/regexp_nfa.c:6758:10
    #12 0x13a2fa0 in vim_regexec_multi /home/wangxinyu/neovim/build/../src/nvim/regexp.c:7335:16
    #13 0x17f3f71 in syn_regexec /home/wangxinyu/neovim/build/../src/nvim/syntax.c:2915:7
    #14 0x17f7d3c in find_endpos /home/wangxinyu/neovim/build/../src/nvim/syntax.c:2634:15
    #15 0x17f0f98 in update_si_end /home/wangxinyu/neovim/build/../src/nvim/syntax.c:2486:3
    #16 0x17f64df in syn_update_ends /home/wangxinyu/neovim/build/../src/nvim/syntax.c:949:9
    #17 0x17b1865 in syn_start_line /home/wangxinyu/neovim/build/../src/nvim/syntax.c:887:5
    #18 0x17a57fa in syntax_start /home/wangxinyu/neovim/build/../src/nvim/syntax.c:553:3
    #19 0x14dc899 in win_line /home/wangxinyu/neovim/build/../src/nvim/screen.c:2236:7
    #20 0x1499bd0 in win_update /home/wangxinyu/neovim/build/../src/nvim/screen.c:1398:15
    #21 0x1479678 in update_screen /home/wangxinyu/neovim/build/../src/nvim/screen.c:497:7
    #22 0x82ed6a in ins_redraw /home/wangxinyu/neovim/build/../src/nvim/edit.c:1418:7
    #23 0x7ea90d in insert_check /home/wangxinyu/neovim/build/../src/nvim/edit.c:613:3
    #24 0x17905fa in state_enter /home/wangxinyu/neovim/build/../src/nvim/state.c:28:35
    #25 0x7efedb in insert_enter /home/wangxinyu/neovim/build/../src/nvim/edit.c:482:5
    #26 0x7e3dd9 in edit /home/wangxinyu/neovim/build/../src/nvim/edit.c:1387:3
    #27 0x11895b6 in invoke_edit /home/wangxinyu/neovim/build/../src/nvim/normal.c:7609:7
    #28 0x118c046 in n_opencmd /home/wangxinyu/neovim/build/../src/nvim/normal.c:7064:7
    #29 0x1159b73 in nv_open /home/wangxinyu/neovim/build/../src/nvim/normal.c:7924:5

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/wangxinyu/neovim/build/../src/nvim/search.c:1605:24 in find_rawstring_end
Shadow bytes around the buggy address:
  0x0c4280137cd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4280137ce0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4280137cf0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4280137d00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c4280137d10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c4280137d20:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4280137d30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4280137d40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4280137d50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4280137d60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c4280137d70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==nvim==20641==ABORTING
```

```
nvim -u NORC
# Alternative for shell-related problems:
# env -i TERM=ansi-256color "$(which nvim)"

```

**Comment by justinmk** [comment]
Might be from one of the vim regex patches. Can you bisect?

**Comment by blueyed** [comment]
> vim -u DEFAULTS 8.1.1186 not crash

Just an aside: it is probably not built with ASAN, is it?  (i.e. it might still have that issue)

**Comment by blueyed** [comment]
To git-bisect this it would be helpful to have a minimal reproducible example, of course.

**Comment by comicfans** [comment]
> > vim -u DEFAULTS 8.1.1186 not crash
> 
> Just an aside: it is probably not built with ASAN, is it? (i.e. it might still have that issue)

Oh I missed this. you're right, vim (from archlinux) is not built with ASAN, it may have similar bug but didn't triggered

**Comment by chrisbra** [comment]
FWIW, I cannot reproduce the ASAN stack with a ASAN compiled Vim master (8.1.2135) and the steps given in the first message.

**Comment by comicfans** [comment]
> FWIW, I cannot reproduce the ASAN stack with a ASAN compiled Vim master (8.1.2135) and the steps given in the first message.

if only type the code snip in an empty rust file, neovim also not trigger the bug. it happens when I type the code inside a particular rust file only. seems neovim is doing completion (or maybe hightlighting)

**Comment by blueyed** [comment]
@comicfans 
Please try to create a minimal reproducible example for us to being able to reproduce / debug it from there then.

**Comment by comicfans** [comment]
@blueyed If I can provide crash core dump, is it enough for directly debug ?

**Comment by blueyed** [comment]
@comicfans 
Sorry for the late reply - that would not help (me) that much.
The best would be if you could git-bisect it, i.e. find out which commit is causing it.

**Comment by comicfans** [comment]
@blueyed  
I'm afraid that I don't have much time to do such bitsect ... sorry about that

**Comment by comicfans** [comment]
@blueyed hello, seems that I've found a reproducible file to crash neovim, but for some reason I can only send a private mail to developers (instead of make it public visible), is it OK ?

**Comment by comicfans** [comment]
hello anybody help ? with some quickly check, I've found that 

```C
find_rawstring_end
```

didn't check line length before accessing it? I'm not sure if vim using c-string internally. anyway following simple patch fix this crash and I'm still checking it worked

```diff
diff --git a/src/nvim/search.c b/src/nvim/search.c
index b105d99d7..4df9ff946 100644
--- a/src/nvim/search.c
+++ b/src/nvim/search.c
@@ -1599,6 +1599,7 @@ static bool find_rawstring_end(char_u *linep, pos_T *startpos, pos_T *endpos)
   char_u *delim_copy;
   size_t delim_len;
   linenr_T lnum;
+  size_t p_len;
 
   for (p = linep + startpos->col + 1; *p && *p != '('; p++) {}
 
@@ -1612,8 +1613,11 @@ static bool find_rawstring_end(char_u *linep, pos_T *startpos, pos_T *endpos)
       if (lnum == endpos->lnum && (colnr_T)(p - line) >= endpos->col) {
         break;
       }
-      if (*p == ')' && p[delim_len + 1] == '"'
-          && STRNCMP(delim_copy, p + 1, delim_len) == 0) {
+
+      p_len = strlen((const char*)p);
+
+      if (*p == ')' && (delim_len + 1 <= p_len) && p[delim_len + 1] == '"'
+          && ( 1 <= p_len ) && STRNCMP(delim_copy, p + 1, delim_len) == 0) {
         found = true;
         break;
       }

```

**Comment by erw7** [comment]
This was fixed in vim by https://github.com/vim/vim/commit/282f9c64e570e832d3eba12950da15c9f1a40f06. The upstream changes have already been merged into neovim by #12713. So this problem should already be fixed in neovim. If it hasn't been fixed, please reopen this.
